### PR TITLE
[FLINK-2363] [docs] First part of internals - complete job execution overview

### PR DESCRIPTION
**Pull request by StephanEwen**
I found that many people do not understand the model behind Flink.

I started drafting/outlining a documentation page that describes end-to-end what happens between the call to `execute()` and the program result being returned or written.

### CODE REVIEW

File: docs/internals/through_stack.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+This page explains what happens when you execute a Flink program (streaming or streaming).
```
Comment (fhueske) [inline docs/internals/through_stack.md]: to stream or not to stream? ;-)

File: docs/internals/through_stack.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+parallel operator instance
+
+execution attempts
+
+
+## TaskManager: Running Tasks
+
+The *{% gh_link flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala "TaskManager" %}*
+is the worker process in Flink. It runs the *{% gh_link flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java "Tasks" %}*,
+which execute a parallel operator instance.
```
Comment (fhueske) [inline docs/internals/through_stack.md]: "a parallel operator instance." -> "parallel operators instances."

File: docs/internals/through_stack.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+The TaskManager itself is an Akka actor and has many utility components, such as:
+
+ - The *{% gh_link flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java "NetworkEnvironment" %}*, which takes care of all data exchanges (streamed and batched) between TaskManagers. Cached data sets in Flink are also cached streams, to the network environment is responsible for that as well.
```
Comment (fhueske) [inline docs/internals/through_stack.md]: Last sentence should be rephrased.

File: docs/internals/through_stack.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+ - The *{% gh_link flink-runtime/src/main/java/org/apache/flink/runtime/memorymanager/MemoryManager.java "MemoryManager" %}*, which governs the memory for sorting, hashing, and in-operator data caching.
+
+ - The *{% gh_link flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/iomanager/IOManager.java "I/O Manager" %}*, which governs the memory for sorting, hashing, and in-operator data caching.
```
Comment (fhueske) [inline docs/internals/through_stack.md]: Copy-Paste from MemoryManager

File: docs/internals/through_stack.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+ 6. The Task invokes the executable code that was configured. This code is usually generic and only differentiates between coarse classes of operations (e.g., {% gh_link flink-runtime/src/main/java/org/apache/flink/runtime/operators/RegularPactTask.java "Plain Batch Task" %} , {% gh_link flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java "Streaming Operator" %}, or {% gh_link flink-runtime/src/main/java/org/apache/flink/runtime/iterative/task/IterationIntermediatePactTask.java "Iterative Batch Task" %}). Internally, these generic operators instantiate the specific operator code (e.g., map task, join task, window reducer, ...) and the user functions and executes them.
+
+ 7. If the Task was deployed before all iof its inputs were available (early deployment), the Task receives updates on those newly available streams.
```
Comment (fhueske) [inline docs/internals/through_stack.md]: "iof" -> "of"

**Comment by fhueske** [comment]
The text for the task manager execution looks good except for some minor things. I did not check the SVG. 

If you want to put it out like that (incomplete sections), I would add a "Work in progress" in the empty sections to make clear that these gaps will be filled.

**Comment by StephanEwen** [comment]
Thanks for the checks. So many typos ;-)

I opened this pull request as WIP and would encourage other committers to jump in. I have been doing miles of internal docs, website, and design documents. Would be great if someone else finished this...

**Comment by fhueske** [comment]
I see, thanks for the effort!
I'll take care of the *Batch API: DataSets to JobGraph* section.

**Comment by StephanEwen** [comment]
I'll fix the typos and update this branch...

**Comment by rmetzger** [comment]
Big +1 to document this properly!


The PR contains headlines without content. I would either remove these headlines before merging or merge the PR once its done. 
(By the coding guidelines, WIP PR's are not allowed anyways)

**Comment by StephanEwen** [comment]
@rmetzger  You are right. This may be different as it is not my personal WIP, but a WIP that I'd like to hand over...

**Comment by fhueske** [comment]
Hi @aljoscha, I haven't started and could do it but you can also give it a shot if you'd like to :smile:

**Comment by aljoscha** [comment]
@fhueske Naa, I just want to see it finished. So I'll probably have to write some other parts myself but it would be cool if you could contribute this part.

I'll open my own branch and then people can do pull requests against that, I will try to write the stuff myself or delegate to others for stuff that I don't know very well.

**Pull request closed without merging**